### DISCUSSION TITLE
Umi4如何自定义一个运行时插件

### DISCUSSION THREAD

**Discussion by Jixiangup** [umijs/umi#10036] [Q&A]
<img width="1457" alt="image" src="https://user-images.githubusercontent.com/75556346/206988386-484700b5-e6f8-4aba-b93c-496c637481b8.png">

```typescript
import {IApi} from "umi";
import {defineConfig} from "@/app";

export default (api: IApi) => {
  api.onStart(() => {
    console.log('启动了....')
  })

  api.describe({
    key:'defineConfig',
    config: {
      schema(joi){
        return joi.object({
          mode: joi.string(),
        })
      }
    }
  })
}
```

插件已经找到了

但是因为加在了`app.ts`所以导致报错
![image](https://user-images.githubusercontent.com/75556346/206988645-0e3d0935-a9a6-4bd6-a8e6-6080a02abec7.png)

**Comment by fz6m** [comment]
可以参考这个包的所有插件写法：https://github.com/umijs/umi/blob/master/packages/plugins/src/access.ts

写在 `${plugin}/index.ts` 下面的会被自动导出，给 `import xx from 'umi'` 用：

https://github.com/umijs/umi/blob/835d353e783923907da05e7349a3cab4651e8338/packages/plugins/src/access.ts#L63-L65

runtime 的需要先写一个 `runtime.tsx` （不一定非要叫这个），再 `addRuntimePlugin` ：

https://github.com/umijs/umi/blob/835d353e783923907da05e7349a3cab4651e8338/packages/plugins/src/access.ts#L26-L28

https://github.com/umijs/umi/blob/835d353e783923907da05e7349a3cab4651e8338/packages/plugins/src/access.ts#L163-L165 

这个 runtime 文件作用和 `app.ts` 等价，可以在里面做运行时操作，比如 `export const rootContainer`

**Reply by Jixiangup** [reply]
你好，很感谢你的回复和提供的帮助，我尝试了你的方法 但是目前还是有点问题我不确定是哪里问题

- index.tsx
```
import {IApi} from "umi";
import {withTmpPath} from "@/utils/withTmpPath";
import Runtime from "./runtime";

export default (api: IApi) => {
  api.describe({
    key: 'Settings',
    config: {
      schema(joi) {
        return joi.string();
      },
    },
    enableBy: api.EnableBy.config,
  });

  // runtime.tsx
  api.writeTmpFile({
    path: 'runtime.tsx',
    content: `
      ${Runtime}
    `,
  })

  api.addRuntimePlugin(() => {
    return [
      withTmpPath({
        api, path: 'runtime.tsx'
      })
    ];
  })
}
```

runtime.tsx

```
export const Runtime: FC = () => {
  return (<div></div>)
}

export default Runtime;
```

- config.ts

```
export default {
  npmClient: 'yarn',
  plugins: [require.resolve('../src/plugins/Settings')],
  // Settings: {
  //
  // }
};
```

**Reply by Jixiangup** [reply]
问题已经解决`index`或者`runtime`文件都不应该使用`tsx`还是使用`ts`

**Reply by Jixiangup** [reply]
目前的问题是`register failed, invalid key Runtime from plugin `因为我直接在`runtime.tsx`返回了`(<div>111</div>)`会导致现在报错 应该如何做返回？

**Reply by Jixiangup** [reply]
解决了 runtime中的组件不export就可以了,因为这已经满足我的需求了 所以我就不深究了。。
